### ISSUE TITLE
todo fails in subtests seem to confuse both yath and prove

### ISSUE THREAD

**Issue by wchristian** [Test-More/test-more#751]
https://github.com/Test-More/Test2-Harness/issues/26
https://rt.cpan.org/Ticket/Display.html?id=119930

**Comment by wchristian** [comment]
This happens due to: https://github.com/Test-More/test-more/blob/77ba3f67eaca979366a2f4df406c37773c13dc86/lib/Test/Builder.pm#L670-L671

Commits that changed those lines:

76406b2e61c3caad59e07ed3d84805bce993b7c8
3b59bd1ddb5e7ab3d155808c3e8e12fa47659604
0f0c7d4d6c1b1953244b96ba4a367c27d85837b7
a32352edaddac34a9e4e693436ad6fb9a075eb02
a3e89956254013bf39108486ac354a2d431cef75
b65ef55649c671ae9bfa116961513fb73115466c

**Comment by schwern** [comment]
tl;dr The special handling of adding a newline to diagnostic output to make the formatting come out right under a traditional harness is broken with subtests + TODO. All that logic should move into the TAP formatter.

After digging into this, I figured out this is a quirk of how subtests handle TODO that doesn't translate well from the old architecture to having a formatter.

Subtests do not inherit the TODO status, I believe this is a feature. This breaks a lot of logic that checks `$self->in_todo` including choosing whether to send the diagnostic output to STDOUT (where prove will hide it) or STDERR (where the user will see it). Test::Builder v0.98 handled this by setting a subtest's diagnostic output handle to be the TODO output handle. Then the subtest is always printing diagnostics to STDOUT despite `$self->in_todo` being false.

The new code uses the TAP formatter. This is good, but it doesn't use the formatter for printing that newline. Instead it does it itself manually getting and printing to the diagnostic filehandle. But the logic setting the diagnostic handle to STDOUT has been removed. So that one newline is going to STDERR while the rest goes to STDOUT.

`Test2::API::Context->failure_diag` has the newline logic, so Test::Builder should probably be changed to use that. In addition, since this is a TAP-only feature, that logic should probably be moved to the TAP formatter.

**Comment by schwern** [comment]
The main problem with doing this appears to be that Test::Builder wants to add a few extra spaces to diagnostics. `#   Failed test` vs `# Failed test`. Test2::Formatter::TAP currently doesn't allow that to be configured, but it could.

**Comment by schwern** [comment]
Work is happening on this in issue/751.

**Comment by exodist** [comment]
@schwern you should also be aware of this: https://github.com/Test-More/test-more/blob/issue/751/lib/Test/Builder/Formatter.pm

**Comment by schwern** [comment]
@exodist Yup, I am. It gets hairy and I'm trying to shave away the special cases.

Test::Builder only wants to use `#   ` for failure diagnostics. All other diagnostics use `# `. The the formatter can't tell the difference between diag events, so Test::Builder has to give it hints. I have that working.

The second problem is pushing the extra newline logic down into the Test2::Formatter::TAP. AFAIK this is behavior we want to retain (rip it out and watch what happens) but it should be happening at the Formatter level. It really should be happening at the harness level, but they don't capture STDERR for technical reasons.

I'm having to teach the TAP formatter when to add a newline and when not to. It needs a model to guess whether the cursor is at the start of a line or not. First order approximation was to have `event_diag` check whether the previous event was also a diagnostic, but that doesn't quite do it. It needs to know if there's been an Ok since the last diagnostic which is where the event duplication of Generic came in.

**Comment by exodist** [comment]
Knowledge dump that may help with this:

Look at Test2::Todo in the Test2-Suite dist. That is a *pure* implementation of the TODO logic.

First off I will point out it has a missing feature, it currently only acts on ::Diag and ::Ok events. It should be updated to use the Event API and act on all events that match.

Test::Builder has the same logic entangled throughout to accomplish the same thing... It is kind of a mess due to backcompat constraints.

So, I think we should step back and figure out a better way to accomplish things.

I think a proper 'TODO' implementation (either in Test2::TODO or Test::Builder, maybe they can share one?) would take events, and if they cause failure they will be wrapped as '::Todo' events which add the 'todo' attribute, and a field that stores the original pass/fail value. In a ::Todo event 'causes_fail' is always false. Similarly any event where the 'diagnostics' attribute is true will be wrapped as a ::TodoDiag event.

These new event types will basically be events that delegate to the original event, but override 'causes_fail' and 'diagnostics' and add any extra methods needed.

At the very least the ::Todo and ::TodoDiag events can be in Test-Simple, and used by both Test::Builder and Test2::Todo.

Once we have the 'pre_filter' wrap the events we should hopefully have all the state we need captured so that this problem becomes a lot less tricky.

Things to consider:
 * Do not break backcompat, existing event types must continue to work, and have the same methods available
 * Test2::Tools::Compare currently recommends writing tests that check the event type, I plan to fix this in my ongoing work to fix Test2::Compare.
 * I know from experience that handling this newline wrong will break nearly half of cpan because some very critical modules use Test::Builder::Tester to inspect TAP output :-(

**Comment by exodist** [comment]
Going back to the basic event API we should probably add 2 methods:

1: 'directive'
2: 'reason'

This matches what 'plan' currently uses. In TAP a directive will be added after the summary, and the reason will be added after the directive. Outside of tap a formatter can do as it pleases with this info.

Neither of these fields will effect 'causes_fail' or 'diagnostics' the event producer will be responsible for making sure an event with a 'todo' directive does not cause fail for instance. Though in this case it is actually the todo pre_filter that handles that as no producer should be paying attention to any kind of TODO state (except Test::Builder, which is the crazy backcompat edge case).

**Comment by exodist** [comment]
*Note, the idea above also makes 'plan' events simpler in the TAP filter as TAP can handle all events with 'directive' and 'reason' the same. print ```"$summary #$directive $reason"``` (obviously conditionals need to be added to not print the directive when there is none).

**Comment by exodist** [comment]
@wchristian Just to double check, this doesn't cause failing tests to report as pass, or passing to report as failed correct? This is just a display/formatting/parsing/etc glitch in both cases?

**Comment by wchristian** [comment]
Correct, it is however a quite severe glitch, as particularly with larger test suites it makes the output completely unreadable.
